Improve csv header parsing for batch importer 1

SaschaPeukert opened the pull request:
This one covers point

SaschaPeukert committed:
Adds option to specify crs in csv header for points

Might need more tests, especially for loading relationships with point properties

henriknyman commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputBatchImportIT.java:
```diff
@@ -263,7 +263,7 @@ private File nodeDataAsFile( List<InputEntity> nodeData ) throws IOException
         {
             // Header
-            println( writer, "id:ID,name,point:Point,date:Date,time:Time,dateTime:DateTime,dateTime2:DateTime,localTime:LocalTime," +
+            println( writer, "id:ID,name,point:Point{WGS-84},date:Date,time:Time,dateTime:DateTime,dateTime2:DateTime,localTime:LocalTime," +
```
I think we could keep a normal `point:Point` without the CRS as well

henriknyman commented on community/csv/src/main/java/org/neo4j/csv/reader/BufferedCharSeeker.java:
```diff
@@ -267,9 +267,9 @@ private static boolean getTrimStringIgnoreErrors( Configuration config )
 
     @Override
-    public <EXTRACTOR extends Extractor<?>> EXTRACTOR extract( Mark mark, EXTRACTOR extractor )
+    public <EXTRACTOR extends Extractor<?>> EXTRACTOR extract( Mark mark, EXTRACTOR extractor, String optionalData )
```
Maybe we could have an overloaded version without the optional parameter that calls this one with `null` to avoid a lot of additional null arguments in the code below.

henriknyman commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/DataFactories.java:
```diff
@@ -427,4 +443,25 @@ else if ( isRecognizedType( typeSpec ) )
+    {
+        String optionalParameter = null;
+        String newTypeSpec = typeSpec;
+        //TODO: This could be done with some clever regex magic
```
Most likely :)

sherfert replied:
`(?<newTypeSpec>.+?)(\{(?<optionalParameter>.*)\})?$`

I find https://regex101.com/ very helpful for this.

henriknyman commented on community/values/src/test/java/org/neo4j/values/storable/PointTest.java:
```diff
@@ -110,6 +111,19 @@ public void shouldBeAbleToParsePointWithUnquotedCrs()
+        }
+        catch ( IllegalArgumentException e )
+        {
+            assertEquals( "Conflicting crs attributes for point found. Header crs states 'wgs-84' while column specifies 'wgs-84-3D'", e.getMessage() );
```
I personally think that it would be useful to have the default in the header and then be able to override it per value. We _could_ even keep the same field format as a Point value to allow for other default values, but I am not sure if we should.

SaschaPeukert committed:
Overloading method to reduce number of passing null to extract calls

SaschaPeukert committed:
Add point value without given crs to import test

SaschaPeukert committed:
Introduce full map based approach for csv header format

Data map entries override the (optional) header map entries

SaschaPeukert committed:
Small test fixes and making checkstyle happy

sherfert commented on community/csv/src/main/java/org/neo4j/csv/reader/Extractors.java:
```diff
@@ -1028,9 +1040,11 @@ public AnyValue value()
-        protected boolean extract0( char[] data, int offset, int length )
+        protected boolean extract0( char[] data, int offset, int length, AnyValue[] optionalData )
         {
+            //TODO: Change DateValue parse to use optionalData
```
So this will happen in a different PR I guess?

SaschaPeukert replied:
Correct!

sherfert commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/csv/DataFactories.java:
```diff
@@ -293,21 +297,26 @@ public boolean isDefined()
+
+            if ( rawHeaderField != null )
+            {
+                String rawHeaderUntilOptions = rawHeaderField.split( "\\{" )[0];
```
Does that mean the header must not contain `{` anywhere before the new options? Can we always guarantee that?

SaschaPeukert replied:
Please correct me if I am wrong but as far as I see it, ```{``` cannot be part of a variable name or a label and is not part of a data type. Thus it should be safe to use it that way.

sherfert replied:
Seems correct to me.

sherfert commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/input/csv/CsvInputTest.java:
```diff
@@ -488,6 +488,27 @@ public void shouldParsePointPropertyValues() throws Exception
+        DataFactory data = data(
+                ":ID,name,point:Point{crs:WGS-84-3D}\n" +
+                        "0,Johan,\" { height :0.01 ,longitude:5, latitude : -4.2 } \"\n" );
+        Iterable<DataFactory> dataIterable = dataIterable( data );
```
Another good test would be with WGS-84 in the header and x and y in the 1st row. Then we should see it chooses WGS-84 instead of Cartesian.

SaschaPeukert replied:
Good idea!

sherfert commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -280,9 +280,32 @@ public static PointValue fromMap( MapValue map )
+        if ( fieldsFromHeader != null )
+        {
+            //It is given that fieldsFromData.length == fieldsFromHeader.length because parseIntoArray produces fixed length arrays
+            // Merge InputFields: Data fields override header fields
```
That should have a JavaDoc (e.g. `fieldsFromHeader must be a value obtained from parseIntoArray` ) and an actual assert statement would also be nice.

henriknyman replied:
Or with if we do the change to `defaultFields` we can just do the for loop from 0 to min(fieldsFromData.length, defaultFields.length).

sherfert reviewed (changes requested):
It looks mostly good to me.

One thing I dislike it that we expose `parseIntoArray`, which is rather an implementation detail. Of course we do not want to repeat parsing the header multiple times. An alternative would be a method `public AugmentedPointParser withHeader(CharSequence header)` where an `AugmentedPointParser` has a method `parse(CharSequence text)`. Don't know if that is better and worth it, though.

henriknyman commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -280,9 +280,32 @@ public static PointValue fromMap( MapValue map )
+        return PointValue.parse( text, null );
+    }
+
+    public static PointValue parse( CharSequence text, AnyValue[] fieldsFromHeader )
```
How about renaming `fieldsFromHeader` to something like `defaultFields` to make it user agnostic?

SaschaPeukert committed:
Include some review feedback

More javadoc
Added a relevant test
Use regex for splitting typeSpec and optionalParameter

henriknyman commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -280,9 +280,32 @@ public static PointValue fromMap( MapValue map )
+        return fromInputFields( fieldsFromData );
+    }
+
+    public static AnyValue[] parseIntoArray( CharSequence text )
```
We could encapsulate this as a `PointValueFields` (static inner) class if we make this public. Then we also know we have the correct length when we pass it in as a `defaultFields` argument.

henriknyman commented on community/values/src/main/java/org/neo4j/values/storable/PointValue.java:
```diff
@@ -280,9 +280,42 @@ public static PointValue fromMap( MapValue map )
+     * @param fieldsFromHeader must be a value obtained from {@link #parseIntoArray(CharSequence)} or null
+     * @return a PointValue instance with information from the {@param fieldsFromHeader} and {@param text}
+     */
+    public static PointValue parse( CharSequence text, AnyValue[] fieldsFromHeader )
```
[Repeating my comment from the outdated diff]
How about renaming `fieldsFromHeader` to something like `defaultFields` to make it user agnostic?

sherfert commented:
It looks good to me. Any further improvements are optional. 

@SaschaPeukert Since I don't know if you wanna include some of Henrik's feedback or not, I pass the honor of merging this over to you.
